substrate/primitives/version/src/lib.rs

```diff
@@ -230,9 +221,406 @@ pub struct RuntimeVersion {
 	/// This number should never decrease.
 	pub transaction_version: u32,
 
-	/// Version of the state implementation used by this runtime.
+	/// Version of the system implementation used by this runtime.
 	/// Use of an incorrect version is consensus breaking.
-	pub state_version: u8,
+	pub system_version: u8,
+}
+
+// Manual implementation in order to sprinkle `stateVersion` at the end for migration purposes
+// after the field was renamed from `state_version` to `system_version`
+#[cfg(feature = "serde")]
+impl serde::Serialize for RuntimeVersion {
```

Comment:
Hmm.... I don't think we need this `serde` boilerplate? Something like this should work:

```rust
use serde::{Serialize, Deserialize};

#[derive(Clone)]
pub struct RuntimeVersion {
    // ...
    pub system_version: u8
}

#[derive(Clone, Serialize, Deserialize)]
struct RuntimeVersionSerde {
    // ...
    system_version: Option<u8>,
    state_version: Option<u8>,
}

impl From<RuntimeVersion> for RuntimeVersionSerde {
    fn from(value: RuntimeVersion) -> Self {
        Self {
            // ...
            system_version: Some(value.system_version),
            state_version: Some(value.system_version),
        }
    }
}

impl From<RuntimeVersionSerde> for RuntimeVersion {
    fn from(value: RuntimeVersionSerde) -> Self {
        Self {
            // ..
            system_version: value.system_version.or(value.state_version)
                .unwrap() // TODO: Properly handle if both/none are there.
        }
    }
}

impl serde::Serialize for RuntimeVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        RuntimeVersionSerde::from(self.clone()).serialize(serializer)
    }
}

impl<'de> serde::Deserialize<'de> for RuntimeVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Ok(RuntimeVersionSerde::deserialize(deserializer)?.into())
    }
}
```

Basically - use an auxiliary struct to do the serialization/deserialization, and then just convert between the real `RuntimeVersion` and the auxiliary one. You still need some boilerplate to copy the fields, but it should be much less than the serde junk and be much easier to maintain.

Reply:
This looks appealing, but `Option` and missing field are not 100% the same for `serde`. How big of an issue is this really?

Reply:
Yeah, just let's take the current code.

Reply:
> This looks appealing, but `Option` and missing field are not 100% the same for `serde`. How big of an issue is this really?

@nazar-pc What do you mean are not the same? What exactly is the problem? You can implement strictly the same behavior as you've currently implemented, just with a lot less code (and more elegant) if you use a wrapper like I've suggested.

I'm fine with taking the existing code *if* it's truly necessary, but if not I'd rather have a simplified version that won't be pain to maintain and won't require you to read a ton of serde boilerplate to understand what it does.

Reply:
Serde is not used just for JSON, so while `Some(u8)` and `u8` encode identically in JSON, I do not know if it is actually the case for all formats that serde supports, so the change is not strictly equivalent in serialization case.

Reply:
> Serde is not used just for JSON, so while `Some(u8)` and `u8` encode identically in JSON, I do not know if it is actually the case for all formats that serde supports, so the change is not strictly equivalent in serialization case.

Not entirely sure if we care (do we?), but, okay, fair enough. In that case you can just use two internal structs - one for serialization with both fields as `u8`, and one for deserialization with them being `Option<u8>`. That would solve this problem and still be less code/boilerplate, no?

So, basically, this (quick copy-paste so might not compile):

```rust
/// Runtime version.
/// This should not be thought of as classic Semver (major/minor/tiny).
/// This triplet have different semantics and mis-interpretation could cause problems.
/// In particular: bug fixes should result in an increment of `spec_version` and possibly
/// `authoring_version`, absolutely not `impl_version` since they change the semantics of the
/// runtime.
#[derive(Clone, PartialEq, Eq, Encode, Default, sp_runtime::RuntimeDebug, TypeInfo)]
pub struct RuntimeVersion {
	/// Identifies the different Substrate runtimes. There'll be at least polkadot and node.
	/// A different on-chain spec_name to that of the native runtime would normally result
	/// in node not attempting to sync or author blocks.
	pub spec_name: RuntimeString,

	/// Name of the implementation of the spec. This is of little consequence for the node
	/// and serves only to differentiate code of different implementation teams. For this
	/// codebase, it will be parity-polkadot. If there were a non-Rust implementation of the
	/// Polkadot runtime (e.g. C++), then it would identify itself with an accordingly different
	/// `impl_name`.
	pub impl_name: RuntimeString,

	/// `authoring_version` is the version of the authorship interface. An authoring node
	/// will not attempt to author blocks unless this is equal to its native runtime.
	pub authoring_version: u32,

	/// Version of the runtime specification.
	///
	/// A full-node will not attempt to use its native runtime in substitute for the on-chain
	/// Wasm runtime unless all of `spec_name`, `spec_version` and `authoring_version` are the same
	/// between Wasm and native.
	///
	/// This number should never decrease.
	pub spec_version: u32,

	/// Version of the implementation of the specification.
	///
	/// Nodes are free to ignore this; it serves only as an indication that the code is different;
	/// as long as the other two versions are the same then while the actual code may be different,
	/// it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are
	/// about the only changes that could be made which would result in only the `impl_version`
	/// changing.
	///
	/// This number can be reverted to `0` after a [`spec_version`](Self::spec_version) bump.
	pub impl_version: u32,

	/// List of supported API "features" along with their versions.
	pub apis: ApisVec,

	/// All existing calls (dispatchables) are fully compatible when this number doesn't change. If
	/// this number changes, then [`spec_version`](Self::spec_version) must change, also.
	///
	/// This number must change when an existing call (pallet index, call index) is changed,
	/// either through an alteration in its user-level semantics, a parameter
	/// added/removed, a parameter type changed, or a call/pallet changing its index. An alteration
	/// of the user level semantics is for example when the call was before `transfer` and now is
	/// `transfer_all`, the semantics of the call changed completely.
	///
	/// Removing a pallet or a call doesn't require a *bump* as long as no pallet or call is put at
	/// the same index. Removing doesn't require a bump as the chain will reject a transaction
	/// referencing this removed call/pallet while decoding and thus, the user isn't at risk to
	/// execute any unknown call. FRAME runtime devs have control over the index of a call/pallet
	/// to prevent that an index gets reused.
	///
	/// Adding a new pallet or call also doesn't require a *bump* as long as they also don't reuse
	/// any previously used index.
	///
	/// This number should never decrease.
	pub transaction_version: u32,

	/// Version of the system implementation used by this runtime.
	/// Use of an incorrect version is consensus breaking.
	pub system_version: u8,
}

#[cfg(feature = "serde")]
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct RuntimeVersionSer {
    spec_name: RuntimeString,
    impl_name: RuntimeString,
    authoring_version: u32,
    spec_version: u32,
    impl_version: u32,
    #[serde(serialize_with = "apis_serialize::serialize")]
    apis: ApisVec,
    transaction_version: u32,
    system_version: u8,
    state_version: u8,
}

#[cfg(feature = "serde")]
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct RuntimeVersionDe {
    spec_name: RuntimeString,
    impl_name: RuntimeString,
    authoring_version: u32,
    spec_version: u32,
    impl_version: u32,
    #[serde(deserialize_with = "apis_serialize::deserialize"]
    apis: ApisVec,
    transaction_version: u32,
    system_version: Option<u8>,
    state_version: Option<u8>,
}

#[cfg(feature = "serde")]
impl serde::Serialize for RuntimeVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        RuntimeVersionSer {
            spec_name: self.spec_name.clone(),
            impl_name: self.impl_name.clone(),
            authoring_version: self.authoring_version,
            spec_version: self.spec_version,
            impl_version: self.impl_version,
            apis: self.apis.clone(),
            transaction_version: self.transaction_version,
            system_version: self.system_version,
            state_version: self.system_version,
        }.serialize(serializer)
    }
}

#[cfg(feature = "serde")]
impl<'de> serde::Deserialize<'de> for RuntimeVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let v = RuntimeVersionDe::deserialize(deserializer)?;
        let system_version = match (v.system_version, v.state_version) {
            (Some(version), None) | (None, Some(version)) => version,
            (Some(system_version), Some(state_version)) => {
                if system_version == state_version {
                    system_version
                } else {
                    return Err(<D::Error as serde::de::Error>::custom(
                        alloc::format!(
                            r#"Duplicated "stateVersion" and "systemVersion" \
                            fields must have the same value, but different values \
                            were provided: {system_version} vs {new_value}"#
                        ),
                    ));
                }
            },
            (None, None) => return Err(<D::Error as serde::de::Error>::missing_field("systemVersion")),
        };
        Ok(Self {
            spec_name: v.spec_name,
            impl_name: v.impl_name,
            authoring_version: v.authoring_version,
            spec_version: v.spec_version,
            impl_version: v.impl_version,
            apis: v.apis,
            transaction_version: v.transaction_version,
            system_version,
        })
    }
}
```

Isn't this much cleaner?

Reply:
Same thing with deserialization: you can deserialize a number into both `Option<u8>` and `u8` from JSON, but I do not think it is guaranteed to be symmetrical that way with other formats. If we do not care then fine, but I was trying to make it actually equivalent and didn't find a clean way to do so without manual implementation.

Reply:
Well, as far as I know the only format we ever explicitly supported was JSON (we just used serde because that's the "standard" way to support JSON), and if we care about formats other than JSON then ideally we should have tests for serialization/deserialization of those formats, otherwise considering how flexible serde is we essentially always risk breaking some unknown code somewhere. And we're already partially breaking compatibility anyway by having it serialize an extra field!

So, to reiterate my policy here: if there's a good reason to keep things complex, sure, go for it, but if it's something that's "just in case" and we don't have any idea at all whether it's necessary, then I just prefer to keep things as simple as possible.

Anyway, honestly, taking a step back: the simplest thing would have been to leave the name `state_version` as `state_version` and just introduce the possibility of it being `2`, and then all of this could have been avoided! We wouldn't have to worry whether this breaks anything, we wouldn't need custom serialization/deserialization, we wouldn't need deprecation warnings, consumers of `RuntimeVersion` that don't expect the extra field wouldn't break (if there are any then they will break now!), we wouldn't have to remove the `state_version` from the serialized `RuntimeVersion` in the future potentially breaking consumers *yet again* (or, alternatively, we wouldn't leave those two fields in there forever - do you have a plan for when to remove the `state_version`?), and this PR would have been already merged long ago. The only advantage of renaming it is, as far as I can see, a slightly nicer name, and a big list of pain and disadvantages to go with it.

The RFC was already accepted by the fellowship, so I **won't** block this, but my opinion is that I would *highly* prefer to just leave the old name as it was, delete all of the new extra code and call it a day.

Anyway, okay, sorry, rant over. If Big Boss @bkchr's fine with this then I'm also fine; I don't like it, but I don't see any hard blockers. :P

Reply:
Although, now that I think about it again, isn't emitting both `state_version` and `system_version` non-compliant with the RFC? The RFC stated that only `system_version` should be there. For `polkadot-sdk` internally we're fine to do whatever we want, but on external community/Polkadot interfaces (like the serialized `RuntimeVersion` struct) what is in the Polkadot spec and in fellowship RFCs takes precedence over what we want, so shouldn't we only emit `system_version` or update the RFC?

Reply:
For SCALE the encoding is as documented in the RFC. For others consuming the type via JSON (which is self descriptive and it is observable that the field name changed), they can continue to use their old code without it breaking.

Reply:
(I'm also not 100% happy with the 400 lines of useless code, but there is no real better solution...)